sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/ReadViaSDF.java

```diff
@@ -0,0 +1,861 @@
+        @Element KafkaSourceDescription kafkaSourceDescription,
+        RestrictionTracker<OffsetRange, Long> tracker,
+        WatermarkEstimator watermarkEstimator,
+        OutputReceiver<KafkaRecord<K, V>> receiver) {
+      // If there is no future work, resume with max timeout and move to the next element.
+      if (((HasProgress) tracker).getProgress().getWorkRemaining() <= 0.0) {
+        return ProcessContinuation.resume().withResumeDelay(KAFKA_POLL_TIMEOUT);
+      }
+      Map<String, Object> updatedConsumerConfig =
+          overrideBootstrapServersConfig(consumerConfig, kafkaSourceDescription);
+      // If there is a timestampPolicyFactory, create the TimestampPolicy for current
+      // TopicPartition.
+      TimestampPolicy timestampPolicy = null;
+      if (timestampPolicyFactory != null) {
+        timestampPolicy =
+            timestampPolicyFactory.createTimestampPolicy(
+                kafkaSourceDescription.getTopicPartition(),
+                Optional.ofNullable(watermarkEstimator.currentWatermark()));
+      }
+      try (Consumer<byte[], byte[]> consumer = consumerFactoryFn.apply(updatedConsumerConfig)) {
+        consumerSpEL.evaluateAssign(
+            consumer, ImmutableList.of(kafkaSourceDescription.getTopicPartition()));
+        long startOffset = tracker.currentRestriction().getFrom();
+        long expectedOffset = startOffset;
+        consumer.seek(kafkaSourceDescription.getTopicPartition(), startOffset);
+        ConsumerRecords<byte[], byte[]> rawRecords = ConsumerRecords.empty();
+
+        try {
+          while (true) {
+            rawRecords = consumer.poll(KAFKA_POLL_TIMEOUT.getMillis());
+            // When there is no records from the current TopicPartition temporarily, self-checkpoint
+            // and move to process the next element.
+            if (rawRecords.isEmpty()) {
+              return ProcessContinuation.resume();
+            }
+            for (ConsumerRecord<byte[], byte[]> rawRecord : rawRecords) {
+              if (!tracker.tryClaim(rawRecord.offset())) {
+                return ProcessContinuation.stop();
+              }
+              KafkaRecord<K, V> kafkaRecord =
+                  new KafkaRecord<>(
+                      rawRecord.topic(),
+                      rawRecord.partition(),
+                      rawRecord.offset(),
+                      consumerSpEL.getRecordTimestamp(rawRecord),
+                      consumerSpEL.getRecordTimestampType(rawRecord),
+                      ConsumerSpEL.hasHeaders() ? rawRecord.headers() : null,
+                      keyDeserializerInstance.deserialize(rawRecord.topic(), rawRecord.key()),
+                      valueDeserializerInstance.deserialize(rawRecord.topic(), rawRecord.value()));
+              int recordSize =
+                  (rawRecord.key() == null ? 0 : rawRecord.key().length)
+                      + (rawRecord.value() == null ? 0 : rawRecord.value().length);
+              avgOffsetGap
```

Comment:
It would be nice if instead we had one map/object that was able to handle both the offset gap and the record size.